### PR TITLE
[dbz tx] Distribute processing across workers

### LINKED ISSUE #11666: [dbz tx] Distribute operator across multiple workers
It's currently on one.  We might not need to do that.  Perhaps something like this (petros's suggestion):
```
Exchange::new(move |decode_result| {
    decode_result
        .value
        .unwrap()
        .iter()
        .nth(transaction_idx)
        .unwrap()
        .unwrap_list()
        .iter()
        .nth(tx_metadata_description.data_transaction_id_idx)
        .unwrap()
        .unwrap_str()
        .hashed()
)
```

### PR DESCRIPTION
**Pull request by cjubb39**
As title.

### Motivation
Fixes #11666 

### Testing

Existing tests should be sufficient

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - just better balancing of work across workers

### CODE REVIEW

File: src/storage/src/render/debezium.rs [hunk 1]
```diff
@@ -159,13 +157,50 @@ where
...
+        .nth(tx_transaction_id_idx)
+    {
+        Some(Datum::String(s)) => s.hashed(),
+        // If there's no transaction_id then we won't be able to use this row so aim to distribute evenly
+        _ => time.hashed(),
+    };
+
+    let data_dist = move |result: &DecodeResult| {
+        // If we can't pull out the transaction_id, it doesn't matter which worker we end up on.  Position seems a good
+        // way to spread around the work as it comes in and is likely uncorrelated to the hash.
```
Comment (guswynn) [inline src/storage/src/render/debezium.rs:182]: we are already distributing the decode for debezium by `partition`, might want to stick with that? https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/decode/mod.rs#L369-L372

File: src/storage/src/render/debezium.rs [hunk 2]
```diff
@@ -159,13 +157,50 @@ where
...
+        let default_hash = result.position.hashed();
+        let value = match result.value.as_ref() {
+            Some(Ok(v)) => v,
+            _ => return default_hash,
+        };
+        let transaction = match value.iter().nth(transaction_idx) {
+            Some(Datum::List(l)) => l,
+            _ => return default_hash,
+        };
+        match transaction.iter().nth(data_transaction_id_idx) {
```
Comment (guswynn) [inline src/storage/src/render/debezium.rs:192]: dont we `unwrap` these later on? the _ case here, and panic on anything but `Null` or `List`, should we match that here?

**Review by guswynn — COMMENTED** [review]
correctness seems fine, @petrosagg can probably confirm!
I have some small nits!

File: src/storage/src/render/debezium.rs (hunk 2 above)
Comment (cjubb39) [reply src/storage/src/render/debezium.rs]: yeah makes sense -- will be more future-proof as any changes to what shapes we allow will require us to fix up here and thus properly distribute messages.

...as much as I don't like the unwraps / panics of types littered around

File: src/storage/src/render/debezium.rs (hunk 1 above)
Comment (guswynn) [reply src/storage/src/render/debezium.rs]: @cjubb39 hmm thats a good point, my understanding was the dedup logic depended on the partition local-ness of data, but I could be remembering wrong, @petrosagg may know...

File: src/storage/src/render/debezium.rs (hunk 2 above)
Comment (guswynn) [reply src/storage/src/render/debezium.rs]: I also value consistency over no unwraps, but definitely possible its worth a cleanup issue here!

File: src/storage/src/render/debezium.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage/src/render/debezium.rs]: The requirement for debezium dedup logic is that the same logical message ends up on the same worker. This means that it is safe to distribute the records on any field that remains the same in the case of repetitions. Practically this means that from `SourceOutput` we can use the `key`, `value`, `partition` fields (or subsets of those) just fine. But `position` and `upstream_time_millis` are not the same when a message gets repeated so we can't use those for debezium processing. 

After writing the above I realize that we can change the distribution function to be based on `value` during decoding and remove the [special case for debezium](https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/decode/mod.rs#L369-L372).

And in the debezium tx operator we should distribute based on transaction id and fallback on the value if that doesn't exist to have good distribution

**Review by petrosagg — APPROVED** [review]
LGTM! See comment about using `value` for better distribution

**Pull request merged**